File: Strivers-A2Z-DSA-Sheet/13. Graphs/3. Topo Sort Problems/javascript/03. Course Scheduler 1.js
/*
Question:
There are numCourses courses labeled 0 to numCourses-1. Given prerequisites[i] = [ai, bi]
meaning you must take bi before ai. Return true if you can finish all courses.

Approach:
- Model as a directed graph with Kahn's algorithm (topological sort).
- If we can process all nodes, no cycle exists → can finish all courses.

Time Complexity: O(N + E)
Space Complexity: O(N + E)
*/

function canFinish(numCourses, prerequisites) {
    let indeg = new Array(numCourses).fill(0);
    let adj = Array.from({ length: numCourses }, () => []);

    for (let it of prerequisites) {
        indeg[it[0]]++;
        adj[it[1]].push(it[0]);
    }

    let queue = [];
    for (let i = 0; i < numCourses; i++) {
        if (indeg[i] === 0) queue.push(i);
    }

    let cnt = 0;
    while (queue.length > 0) {
        let node = queue.shift();
        cnt++;
        for (let v of adj[node]) {
            indeg[v]--;
            if (indeg[v] === 0) queue.push(v);
        }
    }

    return cnt === numCourses;
}

File: Strivers-A2Z-DSA-Sheet/13. Graphs/3. Topo Sort Problems/python/03. Course Scheduler 1.py
"""
Question:
There are numCourses courses. Given prerequisites, return true if you can finish all courses.

Approach:
- Model as a directed graph with Kahn's algorithm (topological sort).
- If we can process all nodes, no cycle → can finish all courses.

Time Complexity: O(N + E)
Space Complexity: O(N + E)
"""

from collections import deque


def can_finish(numCourses, prerequisites):
    indeg = [0] * numCourses
    adj = [[] for _ in range(numCourses)]

    for it in prerequisites:
        indeg[it[0]] += 1
        adj[it[1]].append(it[0])

    queue = deque()
    for i in range(numCourses):
        if indeg[i] == 0:
            queue.append(i)

    cnt = 0
    while queue:
        node = queue.popleft()
        cnt += 1
        for v in adj[node]:
            indeg[v] -= 1
            if indeg[v] == 0:
                queue.append(v)

    return cnt == numCourses

